ee/server/service/software_installers.go

```diff
@@ -1668,38 +1677,102 @@ func (svc *Service) softwareBatchUpload(
-				if err == nil {
-					filename = params["filename"]
+			// check if we already have the installer based on the SHA256
+			teamIDs, err := svc.ds.GetSoftwareInstallerByHash(ctx, p.SHA256)
+			if err != nil {
+				return err
+			}
+
+			var tmID uint
+			if teamID != nil {
+				tmID = *teamID
+			}
+
+			foundInstaller, ok := teamIDs[tmID]
+
+			switch {
+			case ok:
+				installer.StorageID = p.SHA256
+				installer.Extension = foundInstaller.Extension
+				installer.Filename = foundInstaller.Filename
+				installer.Version = foundInstaller.Version
+				installer.Platform = foundInstaller.Platform
+				installer.Source = foundInstaller.Source
+				if foundInstaller.BundleIdentifier != nil {
+					installer.BundleIdentifier = *foundInstaller.BundleIdentifier
+				}
+				installer.Title = foundInstaller.Title
+			case !ok && len(teamIDs) > 0:
+				installer.StorageID = p.SHA256
+				for _, i := range teamIDs {
+					// use the first one we find; which one shouldn't matter because they're all the
+					// same installer bytes
+					installer.Extension = i.Extension
+					installer.Filename = i.Filename
+					installer.Version = i.Version
+					installer.Platform = i.Platform
+					installer.Source = i.Source
+					if i.BundleIdentifier != nil {
+						installer.BundleIdentifier = *i.BundleIdentifier
+					}
+					installer.Title = i.Title
+					break
+				}
+			default:
+				if p.SHA256 != "" && p.URL == "" {
+					// there's a hash specified, but no existing installer with that hash and no way
+					// to download the installer (because the URL is empty).
+					return fmt.Errorf("package not found with hash %s", p.SHA256)
 				}
+				// no existing installer bytes, so we should download it
+				var filename string
+				headers, tfr, err := downloadURLFn(ctx, p.URL)
+				if err != nil {
+					return err
+				}
+
+				installer.InstallerFile = tfr
+
+				// set the filename before adding metadata, as it is used as fallback
+				cdh, ok := headers["Content-Disposition"]
+				if ok && len(cdh) > 0 {
+					_, params, err := mime.ParseMediaType(cdh[0])
+					if err == nil {
+						filename = params["filename"]
+					}
+				}
+
+				installer.Filename = filename
 			}
-			installer.Filename = filename
 
-			ext, err := svc.addMetadataToSoftwarePayload(ctx, installer)
-			if err != nil {
-				_ = tfr.Close() // closing the temp file here since it will not be available after the goroutine completes
-				return err
+			var ext string
+			if installer.InstallerFile != nil {
+				ext, err = svc.addMetadataToSoftwarePayload(ctx, installer)
+				if err != nil {
+					_ = installer.InstallerFile.Close() // closing the temp file here since it will not be available after the goroutine completes
+					return err
+				}
+
+				if p.SHA256 != "" && p.SHA256 != installer.StorageID {
+					// this isn't the specified installer, so return an error
+					return errors.New("downloaded installer hash does not match provided hash for installer")
```

Comment:
Will want to include the URL in the error message so it shows up in GitOps runs, so the admin knows what needs fixing.